aws/resource_aws_waf_rate_based_rule.go

```diff
@@ -0,0 +1,219 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/waf"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsWafRateBasedRule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsWafRateBasedRuleCreate,
+		Read:   resourceAwsWafRateBasedRuleRead,
+		Update: resourceAwsWafRateBasedRuleUpdate,
+		Delete: resourceAwsWafRateBasedRuleDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"metric_name": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateWafMetricName,
+			},
+			"predicates": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"negated": &schema.Schema{
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"data_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								value := v.(string)
+								if len(value) > 128 {
+									errors = append(errors, fmt.Errorf(
+										"%q cannot be longer than 128 characters", k))
+								}
+								return
+							},
+						},
+						"type": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								value := v.(string)
+								if value != "IPMatch" && value != "ByteMatch" && value != "SqlInjectionMatch" && value != "SizeConstraint" && value != "XssMatch" {
+									errors = append(errors, fmt.Errorf(
+										"%q must be one of IPMatch | ByteMatch | SqlInjectionMatch | SizeConstraint | XssMatch", k))
+								}
+								return
+							},
+						},
+					},
+				},
+			},
+			"rate_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
```

Comment:
It seems that this field isn't updatable (based on API reference), shouldn't it be marked as `ForceNew` then?

Reply:
`Rate_key` could be only `IP` now. Why do you think it should be recreated if changed?

Reply:
well, you're right that there's no way to change it, but if the user is unaware they might try and at the moment we'd just ignore the field (in `Update`) and support user's assumption that the field was successfully changed.

If we instead delegate the validation work to `Create` then it will fail and user will also realize during `plan` that this can't be changed.